x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -86,93 +81,54 @@ const normalizeErrorResponse = (errors: RuleActionError[]): NormalizedRuleError[
 
-const getErrorResponseBody = (errors: RuleActionError[], rulesCount: number) => {
-  const errorsCount = errors.length;
-  return {
-    message: errorsCount === rulesCount ? 'Bulk edit failed' : 'Bulk edit partially failed',
-    status_code: 500,
-    attributes: {
-      errors: normalizeErrorResponse(errors).map(({ message, ...error }) => ({
-        ...error,
-        message:
-          message.length > MAX_ERROR_MESSAGE_LENGTH
-            ? `${message.slice(0, MAX_ERROR_MESSAGE_LENGTH - 3)}...`
-            : message,
-      })),
-      rules: {
-        total: rulesCount,
-        failed: errorsCount,
-        succeeded: rulesCount - errorsCount,
-      },
-    },
+const buildBulkResponse = (
+  response: KibanaResponseFactory,
+  fetchRulesOutcome: PromisePoolOutcome<string, RuleAlertType>,
+  bulkActionOutcome: PromisePoolOutcome<RuleAlertType, RuleAlertType | null>
+) => {
+  const errors = [...fetchRulesOutcome.errors, ...bulkActionOutcome.errors];
+  const summary = {
+    failed: errors.length,
+    succeeded: bulkActionOutcome.results.length,
+    total: bulkActionOutcome.results.length + errors.length,
   };
-};
 
-const executeActionAndHandleErrors = async (
-  rule: Rule,
-  action: RuleActionFn
-): Promise<RuleActionResult> => {
-  try {
-    await action(rule);
-  } catch (err) {
-    const { message, statusCode } = transformError(err);
-    return {
-      error: { message, statusCode },
-      rule: { id: rule.id, name: rule.name },
-    };
-  }
-};
-
-const executeBulkAction = async (rules: Rule[], action: RuleActionFn, abortSignal: AbortSignal) =>
-  initPromisePool({
-    concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
-    items: rules,
-    executor: async (rule) => executeActionAndHandleErrors(rule, action),
-    abortSignal,
-  });
-
-const getRulesByIds = async ({
-  ids,
-  rulesClient,
-  isRuleRegistryEnabled,
-  abortSignal,
-}: {
-  ids: string[];
-  rulesClient: RulesClient;
-  isRuleRegistryEnabled: boolean;
-  abortSignal: AbortSignal;
-}) => {
-  const readRulesExecutor = async (id: string) => {
-    try {
-      const rule = await readRules({ id, rulesClient, isRuleRegistryEnabled, ruleId: undefined });
-      if (rule == null) {
-        throw Error('Can`t fetch a rule');
-      }
-      return { rule };
-    } catch (err) {
-      const { message, statusCode } = transformError(err);
-      return {
-        error: { message, statusCode },
-        rule: { id },
-      };
-    }
+  const results = {
+    updated: bulkActionOutcome.results
+      .filter(({ item, result }) => item.id === result?.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    created: bulkActionOutcome.results
+      .filter(({ item, result }) => result != null && result.id !== item.id)
+      .map(({ result }) => result && internalRuleToAPIResponse(result)),
+    deleted: bulkActionOutcome.results
+      .filter(({ result }) => result == null)
+      .map(({ item }) => internalRuleToAPIResponse(item)),
```

Comment:
Shouldn't this logic be based on the bulk action type? Currently, the endpoint can be called only with 1 bulk action type (duplicate, enable, edit, etc); every action can end up with rules either being created or updated or deleted. Am I missing something?

Reply:
we already started to discuss this part of logic here: https://github.com/elastic/kibana/pull/125551#discussion_r807274298
I leaning towards having these properties populated depends on action, rather than map multiple times through results and trying to figure out which result should go where

Reply:
Already explained here: https://github.com/elastic/kibana/pull/125551#discussion_r807667162:

> The current logic is pretty straightforward:
> - We had a rule, now it's null - the rule was deleted
> - We had a rule, now there's a rule with a different id - a new rule was created
> - We had a rule, now there's a rule with the same id - the existing rule was modified
>  
> It doesn't depend on our business logic, so there's no need to tightly couple this with bulk actions whatsoever. Instead, I prefer these pieces to be separated.

Reply:
Sounds good, I'll add the comments with my next PR 👍